src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -308,25 +309,25 @@ public async Task ApplyBatchChangeToWorkspaceMaybe_NoLockAsync(bool useAsync, Fu
                 foreach (var documentId in solutionChanges.DocumentIdsRemoved)
                     Workspace.ClearDocumentData(documentId);
             },
-            onAfterUpdate: (_, _) =>
-            {
-                // Now that the project update has actually applied, we can apply the results of it.
-                // For example saving the state and updating file watchers for added/removed references.
-                ApplyProjectUpdateState(projectUpdateState);
-                if (onAfterUpdate != null)
-                {
-                    onAfterUpdate(projectUpdateState);
-                }
-            },
+            onAfterUpdate: null,
             CancellationToken.None).ConfigureAwait(false);
+
+        // Now that the project update has actually applied, we can apply the results of it.
+        // For example saving the state and updating file watchers for added/removed references.
+        //
+        // Importantly this is not done inside the SetCurrentSolution onAfterUpdate as that
+        // will only run *if* the transformation resulted in a changed solution, but this
+        // must run regardless (it is possible we update maps, but did not end up actually changing the sln object) in the transformation.
```

Comment:
i'm realizing i don't know if the semantics for onAfterUpdate are acceptable in the first place.  HOWEVER.  it hink it's too risky to change.  so i'm approving your approach.